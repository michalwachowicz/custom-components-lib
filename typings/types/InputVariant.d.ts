export type InputVariant = "filled" | "outlined" | "standard";
