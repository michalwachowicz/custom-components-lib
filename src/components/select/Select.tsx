import { useEffect, useId, useRef, useState } from "react";
import * as styles from "./Select.module.scss";
import InputWrapper from "../input/InputWrapper";
import type { InputVariant } from "../../types/InputVariant";
import Menu from "../menu/Menu";

interface SelectProps {
  label: string;
  variant?: InputVariant;
  disabled?: boolean;
  error?: boolean | string;
  options: { value: string; content: string }[];
}

export default function Select({
  label,
  variant = "standard",
  disabled = false,
  error = false,
  options,
}: SelectProps) {
  const id = useId();
  const btnRef = useRef<HTMLButtonElement>(null);
  const menuRef = useRef<HTMLUListElement>(null);

  const [open, setOpen] = useState(false);
  const [focused, setFocused] = useState(false);
  const [value, setValue] = useState({ value: "", display: "" });

  const handleSelect = (value: string, display: string) => {
    setValue({ value, display });
    setOpen(false);
    setFocused(true);
  };

  useEffect(() => {
    if (!open) return;

    function handleClickOutside(event: MouseEvent) {
      if (
        menuRef.current &&
        !menuRef.current.contains(event.target as Node) &&
        !btnRef.current?.contains(event.target as Node)
      ) {
        setOpen(false);
      }
    }

    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [open]);

  return (
    <InputWrapper
      variant={variant}
      focused={focused}
      value={value.value}
      error={error}
      label={label}
      labelElement={<>{label}</>}
      inputElement={
        <>
          <button
            ref={btnRef}
            className={`${styles["select-btn"]} ${
              disabled ? styles["select-btn-disabled"] : ""
            }`}
            disabled={disabled}
            onFocus={() => setFocused(true)}
            onBlur={() => setFocused(false)}
            onClick={() => setOpen(!open)}
            aria-expanded={open}
            aria-haspopup='listbox'
            aria-controls={id}
          >
            {value.display}
          </button>

          {open && (
            <Menu
              ref={menuRef}
              id={id}
              currentValue={value.value}
              options={options}
              onSelect={handleSelect}
            />
          )}
        </>
      }
    />
  );
}
