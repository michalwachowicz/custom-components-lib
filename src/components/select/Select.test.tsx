import { act } from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Select from "./Select";
import { InputVariant } from "../../types/InputVariant";

describe("<Select />", () => {
  const options = [
    { value: "1", content: "Option 1" },
    { value: "2", content: "Option 2" },
  ];

  it("opens menu when button is clicked", async () => {
    render(<Select label='Select' options={options} />);

    const user = userEvent.setup();

    await act(async () => {
      await user.click(screen.getByRole("button"));
    });

    expect(screen.getByRole("listbox")).toBeInTheDocument();
    expect(screen.getByText("Option 1")).toBeInTheDocument();
    expect(screen.getByText("Option 2")).toBeInTheDocument();
  });

  it("does not open menu when disabled", async () => {
    render(<Select label='Select' options={options} disabled />);

    const user = userEvent.setup();

    await act(async () => {
      await user.click(screen.getByRole("button"));
    });

    expect(screen.queryByRole("listbox")).not.toBeInTheDocument();
  });

  it("closes menu when clicking outside", async () => {
    render(<Select label='Select' options={options} />);

    const user = userEvent.setup();

    await act(async () => {
      await user.click(screen.getByRole("button"));
    });

    await act(async () => {
      await user.click(screen.getByRole("button"));
    });

    expect(screen.queryByRole("listbox")).not.toBeInTheDocument();
  });

  it("renders label correctly", () => {
    render(<Select label='Select' options={options} />);
    expect(screen.getByText("Select")).toBeInTheDocument();
  });

  it("renders error message if error prop is a string", () => {
    render(
      <Select label='Select' options={options} error='Incorrect input.' />
    );

    expect(screen.getByTestId("input")).toHaveClass("input-error");
    expect(screen.getByText("Incorrect input.")).toBeInTheDocument();
  });

  it("selects a proper option", async () => {
    render(<Select label='Select' options={options} />);

    const user = userEvent.setup();

    await act(async () => {
      await user.click(screen.getByRole("button"));
    });

    await act(async () => {
      await user.click(screen.getByText("Option 1"));
    });

    expect(screen.queryByRole("listbox")).not.toBeInTheDocument();
    expect(screen.queryByText("Option 2")).not.toBeInTheDocument();
    expect(screen.getByText("Option 1")).toBeInTheDocument();
  });

  it("applies focused class when input is focused", async () => {
    render(<Select label='Name' options={options} />);

    const input = screen.getByRole("button");
    const user = userEvent.setup();
    const focusedClass = "input-focused";

    expect(screen.getByTestId("input")).not.toHaveClass(focusedClass);

    await act(async () => {
      await user.click(input);
    });

    expect(screen.getByTestId("input")).toHaveClass(focusedClass);
  });

  it("shows legend when variant is outlined and focused or not empty", async () => {
    render(
      <div>
        <Select label='Name' variant='outlined' options={options} />
        <div data-testid='outside' />
      </div>
    );

    const input = screen.getByRole("button");
    const user = userEvent.setup();

    expect(screen.queryByTestId("input-fieldset")).not.toBeInTheDocument();

    await act(async () => {
      await user.click(input);
    });

    expect(screen.getByTestId("input-fieldset")).toBeInTheDocument();

    await act(async () => {
      await user.click(screen.getByTestId("outside"));
    });

    expect(screen.queryByTestId("input-fieldset")).not.toBeInTheDocument();
  });

  describe("snapshots", () => {
    const variants = ["standard", "outlined", "filled"];
    const errorStates = [true, false, "error message"];
    const disabledStates = [true, false];

    const testCases = variants.flatMap((variant) =>
      errorStates.flatMap((error) =>
        disabledStates.map((disabled) => [variant, error, disabled])
      )
    );

    test.each(testCases)(
      "renders correctly with variant=%s, error=%s, disabled=%s",
      (variant, error, disabled) => {
        const { asFragment } = render(
          <Select
            label='Name'
            variant={variant as InputVariant}
            error={error}
            disabled={!!disabled}
            options={options}
          />
        );

        expect(asFragment()).toMatchSnapshot();
      }
    );
  });
});
