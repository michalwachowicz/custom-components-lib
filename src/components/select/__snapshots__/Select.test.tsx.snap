// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> snapshots renders correctly with variant=filled, error=error message, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-filled"
    data-testid="input-container"
  >
    <div
      class="input input-filled   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rp»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
    <div
      class="input-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=filled, error=error message, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-filled"
    data-testid="input-container"
  >
    <div
      class="input input-filled   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«ro»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
    <div
      class="input-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=filled, error=false, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-filled"
    data-testid="input-container"
  >
    <div
      class="input input-filled   "
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rn»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=filled, error=false, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-filled"
    data-testid="input-container"
  >
    <div
      class="input input-filled   "
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rm»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=filled, error=true, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-filled"
    data-testid="input-container"
  >
    <div
      class="input input-filled   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rl»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=filled, error=true, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-filled"
    data-testid="input-container"
  >
    <div
      class="input input-filled   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rk»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=outlined, error=error message, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-outlined"
    data-testid="input-container"
  >
    <div
      class="input input-outlined   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rj»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
    <div
      class="input-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=outlined, error=error message, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-outlined"
    data-testid="input-container"
  >
    <div
      class="input input-outlined   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«ri»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
    <div
      class="input-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=outlined, error=false, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-outlined"
    data-testid="input-container"
  >
    <div
      class="input input-outlined   "
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rh»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=outlined, error=false, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-outlined"
    data-testid="input-container"
  >
    <div
      class="input input-outlined   "
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rg»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=outlined, error=true, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-outlined"
    data-testid="input-container"
  >
    <div
      class="input input-outlined   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rf»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=outlined, error=true, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-outlined"
    data-testid="input-container"
  >
    <div
      class="input input-outlined   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«re»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=standard, error=error message, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-standard"
    data-testid="input-container"
  >
    <div
      class="input input-standard   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rd»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
    <div
      class="input-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=standard, error=error message, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-standard"
    data-testid="input-container"
  >
    <div
      class="input input-standard   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rc»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
    <div
      class="input-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=standard, error=false, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-standard"
    data-testid="input-container"
  >
    <div
      class="input input-standard   "
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«rb»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=standard, error=false, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-standard"
    data-testid="input-container"
  >
    <div
      class="input input-standard   "
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«ra»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=standard, error=true, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="input-container-standard"
    data-testid="input-container"
  >
    <div
      class="input input-standard   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«r9»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn "
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Select /> snapshots renders correctly with variant=standard, error=true, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="input-container-standard"
    data-testid="input-container"
  >
    <div
      class="input input-standard   input-error"
      data-testid="input"
    >
      <div
        data-wrapper-label="true"
      >
        Name
      </div>
      <div
        class="input-wrapper"
      >
        <button
          aria-controls="«r8»"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="select-btn select-btn-disabled"
          disabled=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
